src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java

```diff
@@ -260,4 +245,90 @@ private void checkByteOrder(ValueLayout vl) {
+        if (firstVariadicArgIndex != -1) {
+            MemoryLayout[] promotedLayouts = function.argumentLayouts().toArray(MemoryLayout[]::new);
+
+            for (int i = firstVariadicArgIndex; i < promotedLayouts.length; i++) {
+                MemoryLayout variadicLayout = promotedLayouts[i];
+
+                if (variadicLayout instanceof ValueLayout vl) {
+                    if (requiresVariadicIntPromotion(vl)) {
+                        promotedLayouts[i] = JAVA_INT;
+                    } else if (requiresVariadicFloatPromotion(vl)) {
+                        promotedLayouts[i] = JAVA_DOUBLE;
+                    }
+                }
+            }
+
+            function = function.returnLayout()
+                    .map(rl -> FunctionDescriptor.of(rl, promotedLayouts))
+                    .orElseGet(() -> FunctionDescriptor.ofVoid(promotedLayouts));
+        }
+
+        return function;
+    }
+
+    private static MethodHandle unpromoteVariadicArgs(MethodHandle mh, FunctionDescriptor unPromotedDesc, LinkerOptions optionSet) {
+        if (optionSet.isVariadicFunction()) {
+            Class<?>[] unpromotedParams = mh.type().parameterArray();
+            List<MemoryLayout> unpromotedLayouts = unPromotedDesc.argumentLayouts();
+            int numPrefixArgs = downcallNumPrefixArgs(unPromotedDesc, optionSet);
+            assert unpromotedLayouts.size() + numPrefixArgs == unpromotedParams.length;
+
+            for (int i = optionSet.firstVariadicArgIndex(); i < unpromotedLayouts.size(); i++) {
+                MemoryLayout variadicLayout = unpromotedLayouts.get(i);
+                if (variadicLayout instanceof ValueLayout vl) {
+                    if (requiresVariadicPromotion(vl)) {
+                        unpromotedParams[numPrefixArgs + i] = vl.carrier();
+                    }
+                }
+            }
+
+            MethodType unpromotedType = MethodType.methodType(mh.type().returnType(), unpromotedParams);
+            mh = mh.asType(unpromotedType);
+        }
+
+        return mh;
+    }
+
+    private static int downcallNumPrefixArgs(FunctionDescriptor descriptor, LinkerOptions options) {
+        int result = 1; // target address
+        if (SharedUtils.returnsGroupLayout(descriptor)) {
+            result++; // SegmentAllocator
+        }
+        if (options.hasCapturedCallState()) {
+            result++;
+        }
+        return result;
+    }
+
+    // For now all of these are the same for every implementation
+    // in the future we might need to make one of them abstract
+    // and implement on a per-linker basis.
+    private static boolean requiresVariadicIntPromotion(ValueLayout vl) {
+        return vl.carrier() == boolean.class || vl.carrier() == byte.class
+                    || vl.carrier() == char.class || vl.carrier() == short.class;
+    }
+
+    private static boolean requiresVariadicFloatPromotion(ValueLayout vl) {
```

Comment:
Should this be called `requiresVariadicDoublePromotion` ? (In the other method we say `int`, which is the target of the conversion).

Reply:
Yes, I can do that. I was going for 'int' meaning 'integral' and 'float' meaning floating point. But I think your suggestion is clearer.